File: 25 Airline Passengers and Stock Price Time Series/25_time_series.R
#R code for chapter 25 of DSILT: Statistics

setwd("C:/Users/Nick/Documents/Word Documents/Data Science Books/DSILT Stats Code/25 Airline Passengers and Stock Price Time Series")

library(tseries)
library(forecast)

d <- read.csv('AirPassengers.csv', header=T)
d$X <- NULL
plot(d$time, d$value, type='l')

#-------------------------------------------------------------------------------------------------#
#---------------------------------Time Series Decomposition---------------------------------------#
#-------------------------------------------------------------------------------------------------#

#Look at ACF and PACF, and perform Augmented Dickey-Fuller test for stationarity
acf(d$value)
pacf(d$value)
adf.test(d$value)  #Stationary

#Note that as the time series increases in magnitude, seasonal variation increases too.  This is a multiplicative relationship (trend*season*random)
#If the above statement were not true, it would be an additive relationship (trend+season+random)

#Smooth by using a centered moving average (data is monthly, so 12 month trend seems appropriate)
trend <- ma(d$value, order=12, centre=T)
plot(d$time, d$value, type='l')
lines(d$time, trend, col='red')

#Remove the trend: remove multiplicative trends by dividing and additive trends by subtracting
d_detrend <- d$value/trend
plot(d$time, d_detrend, type='l')

#Average the seasonality over the period of the MA trend (it was 12 for this dataset)
mseas <- t(matrix(d_detrend, nrow=12))  #Create a matrix of the time periods, where columns are the periods
season <- colMeans(mseas, na.rm=T)      #Column means are the averaged seasonality for each period
plot(as.ts(rep(season, 12)))

#Remove the seasonality: remove multiplicative seasonality by dividing and additive seasonality by subtracting
d_random <- d$value / (trend*season)
plot(as.ts(d_random))

#Do the same thing using the decompose function (requires a time series object) - this is classical decompostion with MAs
?decompose
dts <- ts(d$value, frequency=12)  #Use 12 as frequency because data is measured monthly so a 12 month period seems reasonable
d_decomp <- decompose(dts, "multiplicative")
d_decomp
plot(d_decomp)

#Do the same thing using LOESS decomposition
?stlm
dts <- ts(d$value, frequency=12)
d_stl_decomp <- stlm(dts, "periodic", allow.multiplicative.trend=T)
d_stl_decomp
d_stl_decomp_ts <- as.data.frame(d_stl_decomp$stl)
par(mfrow=c(4,1))
plot(d_stl_decomp_ts$Data, type='l', ylab='observed')
plot(d_stl_decomp_ts$Trend, type='l', ylab='trend')
plot(d_stl_decomp_ts$Seasonal12, type='l', ylab='seasonal')
plot(d_stl_decomp_ts$Remainder, type='l', ylab='random')
par(mfrow=c(1,1))

#Test for stationarity after seasonality and trend have been removed
adf.test(d_stl_decomp_ts$Remainder)  #Stationary
#Perform first differencing to see if that makes series stationary (don't need to here, but it's good for example)
adf.test(diff(d_stl_decomp_ts$Remainder, differences=1))

#-------------------------------------------------------------------------------------------------#
#------------------------------------------ARIMA--------------------------------------------------#
#-------------------------------------------------------------------------------------------------#

#Find best ARIMA model, using BIC, up to maximum orders (max.p, max.d, max.q for ARIMA(p,d,q) models)
arima_model <- auto.arima(d_stl_decomp_ts$Remainder, max.p=5, max.d=5, max.q=5, ic="bic")
plot(arima_model)
tsdisplay(residuals(arima_model), main="ARIMA(2,0,1) Residuals")
Box.test(residuals(arima_model), type='Ljung-Box')  #No evidence of autocorrelation in the residuals
#Check residuals for normality
qqnorm(residuals(arima_model))
qqline(residuals(arima_model))

#Make future predictions for next 24 months
preds <- forecast(arima_model, h=24)
plot(preds)

#Do the same thing using a holdout test set to determine model accuracy
train <- ts(d[1:(144-24), 'value'], frequency=12)
test <- ts(d[(144-24):144, 'value'], frequency=12)
train_stl_decomp <- stlm(train, 'periodic', allow.multiplicative.trend=T)
test_stl_decomp <- stlm(test, 'periodic', allow.multiplicative.trend=T)
train_stl_decomp_ts <- as.data.frame(train_stl_decomp$stl)
test_stl_decomp_ts <- as.data.frame(test_stl_decomp$stl)
arima_model_train <- auto.arima(train_stl_decomp_ts$Remainder, ic='bic')
#arima_model_train <- arima(train_stl_decomp_ts$Remainder, order=c(2,0,1))
preds_test <- forecast(arima_model_train, h=24)
plot(preds_test, main="Forecasted (blue) vs Actual (red) for Last 24 Months")
lines(seq(120,144), test_stl_decomp_ts$Remainder, col='red')

#Holt-Winters exponential smoothing, similar to MA or LOESS smoothing, assumes trend & seasonality are additive
hw_model <- HoltWinters(dts)
hw_model
plot(hw_model)
tsdisplay(residuals(hw_model), main="Holt-Winters Residuals")
Box.test(residuals(hw_model), type='Ljung-Box')  #There IS evidence of autocorrelation in the residuals
#Check residuals for normality
qqnorm(residuals(hw_model))
qqline(residuals(hw_model))
#Plot forecasts
preds_hw <- forecast(hw_model, h=24)
plot(preds_hw)

#ARIMA model the original series to avoid having to add trend and seasonality back to make forecasts
arima_model_two <- auto.arima(dts, ic='bic')
preds_orig_scale <- forecast(arima_model_two, h=24)
plot(preds_orig_scale)

# #Explore using Fourier transform to determine seasonality
# library(TSA)  #Note that this overrides acf and arima from stats package
# 
# #View the periodogram of the Fourier transform of a series
# p = periodogram(as.ts(d$value))
# 
# #Get a dataframe of the frequencies and their power
# freqdf <- data.frame(freq=p$freq, spec=p$spec)
# freqdf$freq <- 1/freqdf$freq  #Convert frequency to time periods (since data is measured in months, this will be months)
# freqdf <- freqdf[order(-freqdf$spec),]  #Order strongest to weakest
# head(freqdf)
# #Biggest seasonality is 144 months (12 years), which is the entire length of the data -> there is a long term trend
# #Seasonality every 6 years and every 1 year

#-------------------------------------------------------------------------------------------------#
#----------------------------------------ARIMA + GARCH--------------------------------------------#
#-------------------------------------------------------------------------------------------------#

#ARIMA + GARCH
library(quantmod)
library(timeSeries)
library(tseries)
library(forecast)
library(rugarch)

#Created differenced log returns of daily closing prices (Cl) for the S&P500 back to 2000
getSymbols("^GSPC", from="2000-01-01")
spret <- diff(log(Cl(GSPC)))
spret[as.character(head(index(Cl(GSPC)),1))] <- 0

#Fit ARIMA
arima_model <- auto.arima(spret, ic='bic')
arima_model
preds <- forecast(arima_model, h=60)  #Forecast next 60 days
plot(preds)

#Fit ARIMA+GARCH
garch_model <- ugarchspec(mean.model=list(armaOrder=c(2, 0), include.mean=T), 
                          variance.model=list(garchOrder=c(1,1)), 
                          distribution.model='sged')
#Fit the model to all but the last 100 observations
garchfit <- ugarchfit(spec=garch_model, data=spret, solver='hybrid', out.sample=100)
garchfit@fit$coef  #Model coefficients
print(garchfit)

#Plot observed log returns vs predicted log returns - notice that the predicted returns show what happends when volatility clustering is removed
plot(spret, main='Actual vs Fitted (red) Values')
lines(fitted(garchfit), col='red')

#Plot the volatility by itself
plot(garchfit@fit$sigma, type='l', ylab='Volatility')

#Plot the volatility with the observed log returns
par(mfrow=c(2,1))
plot(spret, type='l', main='S&P500 Daily Log Returns')
plot(garchfit, which=3)
par(mfrow=c(1,1))

#Plot forecasts
garch_preds <- ugarchforecast(garchfit, n.ahead=100, n.roll=100, out.sample=100)
plot(garch_preds)

#-------------------------------------------------------------------------------------------------#
#---------------------------Cointegration and Pairs Trading---------------------------------------#
#-------------------------------------------------------------------------------------------------#

#Cointegration != correlation
#Start with 2 that you're fairly certain will be cointegrated
getSymbols("GOOGL", from="2012-05-01", to="2018-05-01")
getSymbols("GOOG", from="2012-05-01", to="2018-05-01")
googlp <- unclass(GOOGL$GOOGL.Adjusted)
googp <- unclass(GOOG$GOOG.Adjusted)
plot(googlp, type='l', col='purple')
lines(googp, type='l', col='orange')
plot(googlp, googp)  #This should be generally linear

#Test for cointegration by regressin each on the other
googlm <- lm(googlp~googp)
googm <- lm(googp~googlp)
adf.test(googlm$residuals, k=1)  #Cannot reject null for stationary series -> nonstationary so cannot use cointegration
adf.test(googm$residuals, k=1)   #Cannot reject null for stationary series -> nonstationary so cannot use cointegration

#To try again for 2 other stocks, let's make a function
coint_test <- function(stock1, stock2) {
  stock1m <- lm(stock1~stock2)
  stock2m <- lm(stock2~stock1)
  adf1 <- adf.test(stock1m$residuals, k=1)
  adf2 <- adf.test(stock2m$residuals, k=1)
  if (adf1$p.value>=0.05 & adf2$p.value>=0.05) {
    return(print('Non-stationary, cannot do cointegration'))
  } else if (adf1$statistic < adf2$statistic) {
    return(print(paste0('Use ', colnames(stock1), ' as the dependent variable in the linear combination.')))
  } else {
    return(print(paste0('Use ', colnames(stock2), ' as the dependent variable in the linear combination.')))
  }
}

#Let's cheat and pick 2 that are known to be cointegrated to see what happens
#We'll also cherry pick a time frame during which these stocks were stationary
getSymbols("EWA", from="2008-02-01", to="2012-02-01")
getSymbols("EWC", from="2008-02-01", to="2012-02-01")
ewap <- unclass(EWA$EWA.Adjusted)
ewcp <- unclass(EWC$EWC.Adjusted)
par(mfrow=c(2,1))
plot(ewap, type='l', col='purple')
plot(ewcp, type='l', col='orange')
par(mfrow=c(1,1))
plot(ewap, ewcp)  #This should be generally linear
coint_test(ewap, ewcp)
ewa_hedge_model <- lm(ewap~ewcp)
summary(ewa_hedge_model)
#Hedge ratio is 0.71

#-------------------------------------------------------------------------------------------------#
#---------------------------------------Benford's Law---------------------------------------------#
#-------------------------------------------------------------------------------------------------#

library(benfordsLaw)

census <-  read.csv("Census_Town&CityPopulations.csv", header=TRUE)
#Isolate town population and remove commas and NAs
x <- as.numeric(sub( ",", "", census$X7_2009))
x <- x[!is.na(x)]

first_digit <- benfordFirstDigit(x)
first_two_digits <- benfordFirstTwoDigit(x)
last_two_digits <- benfordLastTwoDigit(x)
plotFD(first_digit)
plotF2D(first_two_digits)
plotL2D(last_two_digits)

#-------------------------------------------------------------------------------------------------#
#---------------------------------Change Point Detection------------------------------------------#
#-------------------------------------------------------------------------------------------------#

library(ecp)

getSymbols("AAPL", from="2008-02-01", to="2012-02-01")
AAPL <- Cl(AAPL)

#AAPL test
aapl_ecp <- e.divisive(X=AAPL, min.size=120)
plot(AAPL)
abline(v=.index(AAPL)[aapl_ecp$estimates[-c(1, length(aapl_ecp$estimates))]], col='blue', lwd=2)

#Get median and mean closing prices for each segment (area between change points)
aapl_segments <- AAPL
aapl_segments$AAPL.Segment <- aapl_ecp$cluster
aapl_seg_meds <- aggregate(aapl_segments, by=list(aapl_segments$AAPL.Segment), FUN=median)
aapl_seg_means <- aggregate(aapl_segments, by=list(aapl_segments$AAPL.Segment), FUN=mean)
#Find differences between segments as percent change from prior segment
diff(aapl_seg_meds)/aapl_seg_meds*100
diff(aapl_seg_means)/aapl_seg_means*100

File: 25 Airline Passengers and Stock Price Time Series/25_time_series.py
#Python code for chapter 25 DSILT: Statistics

'''
-------------------------------------------------------------------------------
---------------------------Time Series Decomposition---------------------------
-------------------------------------------------------------------------------
'''

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from statsmodels.graphics.tsaplots import plot_acf, plot_pacf
from statsmodels.tsa.stattools import adfuller
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.stats.diagnostic import acorr_ljungbox

d = pd.read_csv('/home/dsilt/Desktop/dsilt-stats-code/25 Airline Passengers and Stock Price Time Series/AirPassengers.csv')

#Fix the dates
months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
d['Month'] = months*12
d['time'] = d['time'].astype('str')
d['time'] = d['time'].apply(lambda x: x[0:4])
d['time'] = d['time']+'-'+d['Month']
d['time'] = d['time'].apply(lambda x: pd.datetime.strptime(x, '%Y-%b'))
d.index = d.time
del d['Month'], d['Unnamed: 0'], d['time']
print(d.head())
print(d.dtypes)

#Convert to series to simplify future commands
dts = pd.Series(d['value'])
#print(dts['1949'])
plt.plot(dts)
plt.show()

#Look at ACF and PACF, and perform Augmented Dickey-Fuller test for stationarity
plot_acf(dts)
plt.show()
plot_acf(dts, lags=30)
plt.show()
plot_pacf(dts)
plt.show()
plot_pacf(dts, lags=30)
plt.show()
adftest = adfuller(dts, autolag='AIC')
print('Augmented Dickey-Fuller Test Results',
      '\nTest Statistic:', adftest[0],
      '\np-value:', adftest[1],
      '\nLags Used:', adftest[2],
      '\nNumber of Observations:', adftest[3])
#Cannot reject the null that series is non-stationary

#Note that as the time series increases in magnitude, seasonal variation increases too.  This is a multiplicative relationship (trend*season*random)
#If the above statement were not true, it would be an additive relationship (trend+season+random)

#Smooth by using a centered moving average (data is monthly, so 12 month trend seems appropriate)
trend = dts.rolling(12).mean()
plt.plot(dts)
plt.plot(trend)
plt.show()

#Remove the trend: remove multiplicative trends by dividing and additive trends by subtracting
dts_detrend = dts/trend
plt.plot(dts_detrend)
plt.show()

#Average the seasonality over the period of the MA trend (it was 12 for this dataset)
mseas = np.asarray(dts_detrend).reshape(12, 12).T  #Create a matrix of the time periods, where columns are the periods
season = np.nanmean(mseas, axis=1)                 #Column means are the averaged seasonality for each period
season = np.asarray(list(season)*12)
plt.plot(season)
plt.show()

#Remove the seasonality: remove multiplicative seasonality by dividing and additive seasonality by subtracting
dts_random = dts / (trend*season)
plt.plot(dts_random)
plt.show()

#Do the same thing using the seasonal_decompose function (requires a pandas object) - this is classical decompostion with MAs
help(seasonal_decompose)
dts_decomp = seasonal_decompose(dts, model='multiplicative')
dts_decomp.plot()
plt.show()

#Do the same thing using LOESS decomposition
#This function was taken from https://github.com/jrmontag/STLDecompose
def stl_decompose(df, period=365, lo_frac=0.6, lo_delta=0.01, allow_multiplicative_trend=False):
    '''Create a seasonal-trend (with Loess, aka "STL") decomposition of observed time series data.
    This implementation is modeled after the ``statsmodels.tsa.seasonal_decompose`` method 
    but substitutes a Lowess regression for a convolution in its trend estimation.
    Defaults to an additive model, Y[t] = T[t] + S[t] + e[t]
    For more details on lo_frac and lo_delta, see: 
    `statsmodels.nonparametric.smoothers_lowess.lowess()`
    Args:
        df (pandas.Dataframe): Time series of observed counts. This DataFrame must be continuous (no 
            gaps or missing data), and include a ``pandas.DatetimeIndex``.  
        period (int, optional): Most significant periodicity in the observed time series, in units of
            1 observation. Ex: to accomodate strong annual periodicity within years of daily 
            observations, ``period=365``. 
        lo_frac (float, optional): Fraction of data to use in fitting Lowess regression. 
        lo_delta (float, optional): Fractional distance within which to use linear-interpolation 
            instead of weighted regression. Using non-zero ``lo_delta`` significantly decreases 
            computation time.
    Returns:
        `statsmodels.tsa.seasonal.DecomposeResult`: An object with DataFrame attributes for the 
            seasonal, trend, and residual components, as well as the average seasonal cycle. 
    '''
    import numpy as np
    import pandas as pd
    from pandas.core.nanops import nanmean as pd_nanmean
    from statsmodels.tsa.seasonal import DecomposeResult
    from statsmodels.tsa.filters._utils import _maybe_get_pandas_wrapper_freq
    import statsmodels.api as sm

    # use some existing pieces of statsmodels    
    lowess = sm.nonparametric.lowess
    _pandas_wrapper, _ = _maybe_get_pandas_wrapper_freq(df)

    # get plain np array
    observed = np.asanyarray(df).squeeze()

    # calc trend, remove from observation
    trend = lowess(observed, [x for x in range(len(observed))], 
                   frac=lo_frac, 
                   delta=lo_delta * len(observed),
                   return_sorted=False)
    detrended = observed - trend

    # calc one-period seasonality, remove tiled array from detrended
    period_averages = np.array([pd_nanmean(detrended[i::period]) for i in range(period)])
    # 0-center the period avgs
    period_averages -= np.mean(period_averages)
    seasonal = np.tile(period_averages, len(observed) // period + 1)[:len(observed)]    
    resid = detrended - seasonal

    # convert the arrays back to appropriate dataframes, stuff them back into 
    #  the statsmodel object
    results = list(map(_pandas_wrapper, [seasonal, trend, resid, observed]))    
    dr = DecomposeResult(seasonal=results[0],
                         trend=results[1],
                         resid=results[2], 
                         observed=results[3],
                         period_averages=period_averages)
    return dr

stl = stl_decompose(dts, period=12, allow_multiplicative_trend=False)
stl.plot()
plt.show()

#Test for stationarity after seasonality and trend have been removed
adftest = adfuller(stl.resid, autolag='AIC')
print('Augmented Dickey-Fuller Test Results',
      '\nTest Statistic:', adftest[0],
      '\np-value:', adftest[1],
      '\nLags Used:', adftest[2],
      '\nNumber of Observations:', adftest[3])
#Cannot reject the null that series is non-stationary

#Define differencing function, this was stolen from: https://machinelearningmastery.com/difference-time-series-dataset-python/
def diff(dataset, differences=1):
	diff = list()
	for i in range(differences, len(dataset)):
		value = dataset[i] - dataset[i - differences]
		diff.append(value)
	return pd.Series(diff)

#Perform first differencing to see if that makes series stationary
adftest = adfuller(diff(stl.resid, differences=1), autolag='AIC')
print('Augmented Dickey-Fuller Test Results',
      '\nTest Statistic:', adftest[0],
      '\np-value:', adftest[1],
      '\nLags Used:', adftest[2],
      '\nNumber of Observations:', adftest[3])
#Can reject the null that series is non-stationary, the series is difference stationary

'''
-------------------------------------------------------------------------------
------------------------------------ARIMA--------------------------------------
-------------------------------------------------------------------------------
'''

#Replicate the auto.arima function from R - it uses grid search to opmtimize AIC or BIC
def auto_arima(ts, max_p=5, max_d=5, max_q=5, ic='bic'):
    from statsmodels.tsa.arima_model import ARIMA, ARIMAResults
    import warnings
    with warnings.catch_warnings():
        warnings.simplefilter('ignore')  #Ignores all convergence warnings
        best_ic = 1e24
        best_params = (0,0,0)
        for p in range(max_p):
            for d in range(max_d):
                for q in range(max_q):
                    try:
                        model = ARIMA(ts.astype('float32'), order=(p+1, d+1, q+1))
                        fitted = model.fit()
                        if ic=='aic' and fitted.aic < best_ic:
                            best_ic = fitted.aic
                            best_params = (p+1, d+1, q+1)
                            best_model = fitted
                        elif ic=='bic' and fitted.bic < best_ic:
                            best_ic = fitted.bic
                            best_params = (p+1, d+1, q+1)
                            best_model = fitted
                    except:
                        continue
    return best_model, best_params, best_ic

arima_model, arima_params, arima_ic = auto_arima(dts, max_p=7, max_d=2, max_q=5, ic='bic')
plt.plot(arima_model.fittedvalues)
plt.title('ARIMA Fitted Values')
plt.show()
plt.plot(arima_model.resid)
plt.title('ARIMA Redisuals')
plt.show()
print('Best ARIMA model had params', arima_params, 'and BIC', arima_ic)
#p-values are cumulative for this test, so restrict lags or get spurious significance in distant lags
lag_test_stats, lag_p_vals = acorr_ljungbox(arima_model.resid, lags=10, boxpierce=False)  

#Make future predictions for next 24 months
preds = arima_model.forecast(steps=24) #Returns tuple of preds, stderr, conf_int
plt.plot(np.array(dts))
plt.plot(np.array(range(144, 144+24)), preds[0])  #Shift preds x-axis bc there are 144 obs in original data
plt.title('ARIMA Predictions for Next 24 Months')
plt.show()

'''
-------------------------------------------------------------------------------
----------------------------------ARIMA+GARCH----------------------------------
-------------------------------------------------------------------------------
'''

import datetime as dt
import numpy as np
import pandas as pd
pd.core.common.is_list_like = pd.api.types.is_list_like  #Fixes import error for datareader
import pandas_datareader.data as web
import matplotlib.pyplot as plt
import statsmodels.tsa.api as smt
from arch import arch_model
from statsmodels.stats.diagnostic import acorr_ljungbox


start_dt = dt.datetime(2012, 1, 1)
end_dt = dt.datetime(2016, 1, 1)

d = web.DataReader('SPY', 'morningstar', start_dt, end_dt)
dts = pd.Series(d['Close'])
dts.index= d.index.levels[1]
log_returns = pd.Series(np.log(dts/dts.shift(1)).dropna())
#Note that log returns = percent change
#print(d.pct_change().dropna())
log_returns.plot()
plt.show()
dts.plot()
plt.show()


def auto_arima(ts, max_p=5, max_d=5, max_q=5, ic='bic'):
    from statsmodels.tsa.arima_model import ARIMA, ARIMAResults
    import warnings
    with warnings.catch_warnings():
        warnings.simplefilter('ignore')  #Ignores all convergence warnings
        best_ic = 1e24
        best_params = (0,0,0)
        for p in range(max_p):
            for d in range(max_d):
                for q in range(max_q):
                    try:
                        model = ARIMA(ts.astype('float32'), order=(p+1, d+1, q+1))
                        fitted = model.fit()
                        if ic=='aic' and fitted.aic < best_ic:
                            best_ic = fitted.aic
                            best_params = (p+1, d+1, q+1)
                            best_model = fitted
                        elif ic=='bic' and fitted.bic < best_ic:
                            best_ic = fitted.bic
                            best_params = (p+1, d+1, q+1)
                            best_model = fitted
                    except:
                        continue
    return best_model, best_params, best_ic


arima_model, arima_params, arima_ic = auto_arima(log_returns, max_p=7, max_d=2, max_q=5, ic='bic')
plt.plot(arima_model.fittedvalues)
plt.title('ARIMA Fitted Values')
plt.show()
plt.plot(arima_model.resid)
plt.title('ARIMA Redisuals')
plt.show()
print('Best ARIMA model had params', arima_params, 'and BIC', arima_ic)
#p-values are cumulative for this test, so restrict lags or get spurious significance in distant lags
lag_test_stats, lag_p_vals = acorr_ljungbox(arima_model.resid, lags=10, boxpierce=False)  

#Make future predictions for next 60 days
preds = arima_model.forecast(steps=60) #Returns tuple of preds, stderr, conf_int
plt.plot(np.array(log_returns))
plt.plot(np.array(range(len(dts), len(dts)+60)), preds[0])  #Shift preds x-axis
plt.title('ARIMA Predictions for Next 60 Days')
plt.show()

#Now fit a GARCH model, using the parameters of the best fitting ARIMA model
garch_model = arch_model(log_returns, p=arima_params[0], o=arima_params[1], q=arima_params[2], dist='StudentsT')
garch_fitted = garch_model.fit()
print(garch_fitted.summary())
plt.plot(garch_fitted.resid)
plt.title('GARCH Redisuals')
plt.show()

plot_acf(garch_fitted.resid, lags=30)
plt.show()
plot_pacf(garch_fitted.resid, lags=30)
plt.show()

#View predicted volatility
garch_fitted.hedgehog_plot(horizon=60)
plt.show()

'''
-------------------------------------------------------------------------------
------------------Pairs Trading Using Cointegration----------------------------
-------------------------------------------------------------------------------
'''

'''
Pairs trading uses linear regression to find cointegrated, stationary pairs
Regression finds the mean hedge ratio over a time period.  If the hedge ratio 
changes, then it must be smoothed.  The Kalman filter is one way to smooth it:
https://www.quantstart.com/articles/Dynamic-Hedge-Ratio-Between-ETF-Pairs-Using-the-Kalman-Filter
Pairs trading hedges two stocks and relies on the ADF test to look for 
stationarity.  Hedging a basket of stocks relies on the Johansson test for 
stationarity:
https://quant.stackexchange.com/questions/14327/need-help-on-cointegration
For basket trading, linear regression can again be used to find the hedge 
ratios, and the Kalman filter can again be used to smooth dynamic hedge ratios.
Hedge ratios can be assumed to be drawn from normal distributions, so z-scores
or confidence intervals can bound the hedge ratio.  The hedge ratio is mean 
reverting by definition (stationary time series are mean reverting), so any 
time it diverges from the mean, trading can be done.
'''

#Let's cheat and pick 2 that are known to be cointegrated to see what happens
#We'll also cherry pick a time frame during which these stocks were stationary
start_dt = dt.datetime(2008, 2, 1)
end_dt = dt.datetime(2012, 2, 1)

ewap = web.DataReader('EWA', 'morningstar', start_dt, end_dt)
ewcp = web.DataReader('EWC', 'morningstar', start_dt, end_dt)

ewapts = pd.Series(ewap['Close'])
ewcpts = pd.Series(ewcp['Close'])
ewapts.index = ewap.index.levels[1]
ewcpts.index = ewcp.index.levels[1]

plt.plot(ewapts)
plt.plot(ewcpts)
plt.show()

#Define function to check for cointegration using linear regression
def coint_test(stockA, stockB):
    from sklearn.linear_model import LinearRegression
    from statsmodels.tsa.stattools import adfuller
    stockA = stockA.values.reshape(-1, 1)
    stockB = stockB.values.reshape(-1, 1)
    stockAm = LinearRegression().fit(stockB, stockA)
    stockBm = LinearRegression().fit(stockA, stockB)
    stockAm_resid = stockA-stockAm.predict(stockB)
    stockBm_resid = stockB-stockBm.predict(stockA)
    adfA = adfuller(stockAm_resid.reshape(stockAm_resid.shape[0],), autolag='AIC')
    adfB = adfuller(stockBm_resid.reshape(stockBm_resid.shape[0],), autolag='AIC')
    if (adfA[1] >= 0.05 and adfB[1] >= 0.05):
        print('Non-stationary, cannot do cointegration')
    elif adfA[0] < adfB[0]:
        print('Use the first input as the dependent variable in the linear combination.')
    else:
        print('Use the second input as the dependent variable in the linear combination.')

coint_test(ewapts, ewcpts)
from sklearn.linear_model import LinearRegression
ewa_hedge_model = LinearRegression().fit(ewcpts.values.reshape(-1, 1), ewapts.values.reshape(-1, 1))
print('Hedge ratio:', ewa_hedge_model.coef_)

#Use a Kalman filter to estimate the dynamic hedge ratio
from pykalman import KalmanFilter
def dynamic_regression_w_kalman(x, y):
    #w_t = covariance matrix of the noise term - decrease it to produce smoother result (the smaller, the less sensitive KF is to true values)
    w_t = 1e-5
    #A_t or state transition matrix is the matrix of beta coefficients for the regression - this will start off as a matrix of diagonal noise terms to simulate random walk
    state_transition_matrix = w_t/(1-w_t)*np.eye(2)
    #x_t or the observed matrix is the matrix of values of the independent variable on the left and the dependent variable on the right: [x, y] - initialize dependent vars as ones
    observation_matrix = np.vstack([x, np.ones(y.shape)]).T[:, np.newaxis]
    
    kf = KalmanFilter(n_dim_obs=1, n_dim_state=2,
                      initial_state_mean=np.zeros(2), 
                      initial_state_covariance=np.ones((2, 2)),
                      transition_matrices=np.eye(2), 
                      transition_covariance=state_transition_matrix,
                      observation_matrices=observation_matrix, 
                      observation_covariance=1.0)
    
    return kf.filter(y.values)

state_means, state_covariances = dynamic_regression_w_kalman(ewcpts, ewapts)
dynamic_reg_params = pd.DataFrame({
        'slope': state_means[:, 0],
        'intercept': state_means[:, 1]
        }, index=ewapts.index)
dynamic_reg_params.plot()
plt.title('Dynamic Hedge Ratio of EWA to EWC')
plt.show()

'''
-------------------------------------------------------------------------------
--------------------------Change Point Detection-------------------------------
-------------------------------------------------------------------------------
'''

import datetime as dt
import numpy as np
import pandas as pd
pd.core.common.is_list_like = pd.api.types.is_list_like  #Fixes import error for datareader
import pandas_datareader.data as web
import matplotlib.pyplot as plt
import ruptures as rpt

start_dt = dt.datetime(2008, 2, 1)
end_dt = dt.datetime(2017, 2, 1)

aapl = web.DataReader('AAPL', 'morningstar', start_dt, end_dt)

aaplpts = pd.Series(aapl['Close'])
aaplpts.index = aapl.index.levels[1]

plt.plot(aaplpts)
plt.show()

cp_model = rpt.Pelt(model='rbf', min_size=2).fit(aaplpts.values)
cp_preds = cp_model.predict(pen=3)  #Smaller penalty -> more change points
cp_preds = cp_preds[:-1]  #Drop the final change point, as it is at the end of the series

plt.plot(aaplpts)
plt.title('Change Points in AAPL Price')
for cp in cp_preds:
    plt.axvline(x=aaplpts.index[cp], color='purple', linestyle='--')
plt.show()

print('Change points occurred at:')
for cp in cp_preds:
    print(aaplpts.index[cp])

